src/opencl/mod.rs

```diff
@@ -59,42 +64,83 @@ mod ffi {
+/// buffer is created to write the contents of piece data. Using pinned memory without creating
+/// another buffer during the encode process and writing the contents of piece data to the device
+/// buffer created initially when allocating pinned memory is possible, without any wasted device
+/// memory, but it results in an overall slower encoding time since there's extra memory allocation.
+/// See: https://stackoverflow.com/questions/42011504/why-does-clcreatebuffer-with-cl-mem-alloc-host-ptr-use-discrete-device-memory
+///
+/// Therefore we can only use around 1/3 of the available amount of device memory at maximum.
 pub fn pinned_memory_alloc(instance: *const u8, size: usize) -> Result<Vec<u8>, OpenCLEncodeError> {
     let mut return_code: i32 = 0;
 
     let pointer = unsafe { ffi::sloth256_189_pinned_alloc(instance, size, &mut return_code) };
 
     if return_code != 0 {
-        return Err(OpenCLEncodeError::OpenCLError(return_code));
+        return Err(OpenCLEncodeError::OpenCLError(get_opencl_error_string(
+            return_code,
+        )));
     }
 
     let pointer_as_vec = unsafe { Vec::from_raw_parts(pointer, size, size) };
 
     return Ok(pointer_as_vec);
 }
 
-/// Free the pinned memory allocated with pinned_memory_alloc function above. A std::mem::forget({vector})
-/// call is required since the vector in Rust will be using memory freed in C.
+/// Free the pinned memory allocated with pinned_memory_alloc function above.
+/// A std::mem::forget({vector}) call is required since the vector in Rust will
+/// be using memory freed in C.
 /// Call this function before calling the cleanup function
 pub fn pinned_memory_free(instances: *const u8) -> Result<(), OpenCLEncodeError> {
     let return_code = unsafe { ffi::sloth256_189_pinned_free(instances) };
 
     if return_code != 0 {
-        return Err(OpenCLEncodeError::OpenCLError(return_code));
+        return Err(OpenCLEncodeError::OpenCLError(get_opencl_error_string(
+            return_code,
+        )));
     }
 
     return Ok(());
 }
 
-/// Initializes the encode kernel for a given device. In essence performs the tasks that must be done only once at the start.
-/// Returns a pointer to the initialized instance which should be passed to the encode and cleanup functions.
+/// Initializes the encode kernel for a given device. In essence performs
+/// the tasks that must be done only once at the start.
+/// Returns a pointer to the initialized instance which should be passed
+/// to the encode and cleanup functions.
 /// Calling this once at the start is sufficient.
 pub fn initialize() -> Result<*const u8, OpenCLEncodeError> {
+    let encode_cl = include_bytes!("encode.cl");
+    let nvidia_specific_cl = include_bytes!("nvidia_specific.cl");
+    let mod256_189_cu = include_bytes!("mod256-189.cu");
+    let non_nvidia_cl = include_bytes!("non_nvidia.cl");
+
+    let encode_cl_str = String::from_utf8_lossy(encode_cl).to_string() + &"\0".to_string();
```

Comment:
I will refactor this, but FYI https://doc.rust-lang.org/stable/std/ffi/struct.CString.html exists and `format!()` is more commonly used instead of string concatenation will will do more allocations.

So I'd write this instead:
```rust
    let encode_cl_cstr = CString::new(include_str!("encode.cl")).expect("File doesn't contain 0 bytes");
```

When `CString::new` will become `const`, the whole thing will be processed at compile time and with 0 byte at the end.